File: Computer_Science/Algorithm/Learning/Algorithms/回溯/全排列/code/46.全排列.cpp
/*
 * @lc app=leetcode.cn id=46 lang=cpp
 *
 * [46] 全排列
 */

#include <cstring>
#include <vector>
#include <unordered_set>
using namespace std;
// @lc code=start
class Solution {
public:
    // 判断这个数字在不在路径中
    vector<int> path;
    vector<bool> on_path;
    vector<vector<int>> ans;

    void dfs(vector<int>& nums, int i) {
        int n = nums.size();
        if (i == n) {
            ans.emplace_back(path);
            return;
        }

        for (int j = 0; j < n; j++) {
            if (!on_path[j]) {
                on_path[j] = true;
                path[i] = nums[j];
                dfs(nums, i + 1);
                on_path[j] = false;
            }
        } 
    }


    vector<vector<int>> permute(vector<int>& nums) {
        path.resize(nums.size());
        on_path.resize(nums.size());
        dfs(nums, 0);
        return ans;
    }
};
// @lc code=end

File: Computer_Science/Algorithm/leetcode/hot100/46.全排列.java
/*
 * @lc app=leetcode.cn id=46 lang=java
 *
 * [46] 全排列
 */

// @lc code=start

import java.util.ArrayList;
import java.util.List;


class Solution {
    
    private List<List<Integer>> ans = new ArrayList<>();


    // 枚举选择哪个
    public List<List<Integer>> permute(int[] nums) {
        List<Integer> list = new ArrayList<>(nums.length);
        boolean[] onPath = new boolean[nums.length];
        dfs(0, nums, onPath, list);
        return ans;
    }

    private void dfs(int i, int[] nums, boolean[] onPath, List<Integer> list) {
        if (i == nums.length) {
            ans.add(new ArrayList<>(list));
            return;
        }
        for (int j = 0; j < nums.length; j++) {
            if(!onPath[j]) { // 这个节点没有访问过
                onPath[j] = true;
                list.add(nums[j]);
                dfs(i+1, nums, onPath, list);
                list.remove(list.size()-1);
                onPath[j] = false;
            }
        }
    }
}
// @lc code=end

